openTEPES/openTEPES.py

```diff
@@ -17,74 +17,75 @@
-    idxDict['y'  ] = 1
-
-    #%% model declaration
-    mTEPES = ConcreteModel('Open Generation, Storage, and Transmission Operation and Expansion Planning Model with RES and ESS (openTEPES) - Version 4.17.8 - October 17, 2024')
-    print(                 'Open Generation, Storage, and Transmission Operation and Expansion Planning Model with RES and ESS (openTEPES) - Version 4.17.8 - October 17, 2024', file=open(_path+'/openTEPES_version_'+CaseName+'.log','w'))
-
-    pIndOutputResults = [j for i,j in idxDict.items() if i == pIndOutputResults][0]
-    pIndLogConsole    = [j for i,j in idxDict.items() if i == pIndLogConsole   ][0]
-
-    # introduce cycle flow formulations in DC and AC load flow
-    pIndCycleFlow = 0
-
-    # Define sets and parameters
-    InputData(DirName, CaseName, mTEPES, pIndLogConsole)
-
-    # Define variables
-    SettingUpVariables(mTEPES, mTEPES)
-
-    # first/last stage
-    FirstST = 0
-    for st in mTEPES.st:
-        if FirstST == 0:
-            FirstST = 1
-            mTEPES.First_st = st
-        mTEPES.Last_st = st
-
-    # objective function and investment constraints
-    TotalObjectiveFunction    (mTEPES, mTEPES, pIndLogConsole)
-    InvestmentModelFormulation(mTEPES, mTEPES, pIndLogConsole)
-
-    # initialize parameter for dual variables
-    mTEPES.pDuals = {}
-
-    # control for not repeating the stages in case of several ones
-    mTEPES.NoRepetition = 0
-
-    # initialize the set of load levels up to the current stage
-    mTEPES.na = Set(initialize=[])
 
-    # iterative model formulation for each stage of a year
-    for p,sc,st in mTEPES.ps*mTEPES.stt:
+@dataclass
+class Simulation:
+    mTEPES: ConcreteModel
+    DirName: str
+    CaseName: str
+    SolverName: str
+    pIndOutputResults: int # TODO: make this bool instead
+    pIndLogConsole: int # TODO: make this bool instead
+    pIndCycleFlow: int = 0 # introduce cycle flow formulations in DC and AC load flow
+
+    def __post_init__(self):
+        #%% replacing string values by numerical values
+        idxDict        = dict()
+        idxDict[False] = 0
+        idxDict[0    ] = 0
+        idxDict[0.0  ] = 0
+        idxDict[0.0  ] = 0
+        idxDict['No' ] = 0
+        idxDict['NO' ] = 0
+        idxDict['no' ] = 0
+        idxDict['N'  ] = 0
+        idxDict['n'  ] = 0
+        idxDict[True ] = 1
+        idxDict['Yes'] = 1
+        idxDict['YES'] = 1
+        idxDict['yes'] = 1
+        idxDict['Y'  ] = 1
+        idxDict['y'  ] = 1
+
+        self.pIndLogConsole = idxDict[self.pIndLogConsole]
+        self.pIndOutputResults = idxDict[self.pIndOutputResults]
+        self.path = os.path.join(self.DirName, self.CaseName)
+
+    def input_data(self):
+        # Define sets and parameters
+        InputData(self.DirName, self.CaseName, self.mTEPES, self.pIndLogConsole)
+
+    def define_variables(self):
+        # Define variables
+        SettingUpVariables(self.mTEPES, self.mTEPES)
+
+    def setup_first_last_stage(self):
```

Comment:
The set `mTEPES.st`represents different stages, typically structured as `"st" + number`. However, this set is unordered, which means that the initial and final stages may vary with each execution, even when using identical input data.

Reply:
We changed everything to an ordered set in https://github.com/IIT-EnergySystemModels/openTEPES/pull/66. So the sorting of pyomo should handle this correctly here, right?

Reply:
I think we removed both `ordered=True` and `ordered=False`. Perhaps, it might be more effective to determine the first and last stages based on the position of each stage's initial time step within the overall collection of time steps considered in the model.

Reply:
To clarify, the default in pyomo is ordered=True, so when we removed the keyword argument we made everything ordered.

Reply:
Thanks for the clarification @kdheepak. If the default in Pyomo is indeed ordered=True, then it should handle the stage ordering as expected. Let me know if you think we should run additional tests to verify this.